ca/ca.go

```diff
@@ -646,16 +628,71 @@ func (ca *certificateAuthorityImpl) issuePrecertificateInner(ctx context.Context
+// verifyTBSCertIsDeterministic verifies that x509.CreateCertificate signing
 // operation is deterministic and produced identical DER bytes between the given
-// lintCert and leafCert. If this fails it's mississuance, but it's better to
-// know about the problem sooner than later.
-func verifyTBSCertificateDeterminism(lintCert *x509.Certificate, leafCert *x509.Certificate) error {
-	if core.IsAnyNilOrZero(lintCert, leafCert) {
-		return fmt.Errorf("lintCertBytes and certBytes were nil")
-	}
-	if !bytes.Equal(lintCert.RawTBSCertificate, leafCert.RawTBSCertificate) {
-		return fmt.Errorf("mismatch between lintCert and leafCert RawTBSCertificate DER bytes: \"%x\" != \"%x\"", lintCert.RawTBSCertificate, leafCert.RawTBSCertificate)
+// lint certificate and leaf certificate. If the DER byte equality check fails
+// it's mississuance, but it's better to know about the problem sooner than
+// later. The caller is responsible for passing the appropriate valid
+// certificate bytes in the correct position.
+func tbsCertIsDeterministic(lintCertBytes []byte, leafCertBytes []byte) error {
+	if core.IsAnyNilOrZero(lintCertBytes, leafCertBytes) {
+		return fmt.Errorf("lintCertBytes of leafCertBytes were nil")
+	}
+
+	// parseTBSCert is a partial copy of //crypto/x509/parser.go to extract the
+	// RawTBSCertificate field from given DER bytes. It the RawTBSCertificate
+	// field bytes or an error if the given bytes cannot be parsed. This is far
+	// more performant than parsing the entire *Certificate structure with
+	// x509.ParseCertificate().
+	parseTBSCert := func(inputDERBytes *[]byte) ([]byte, error) {
+		if inputDERBytes == nil {
+			return nil, errors.New("inputDERBytes was nil")
+		}
+
+		input := cryptobyte.String(*inputDERBytes)
+		// we read the SEQUENCE including length and tag bytes so that
+		// we can populate Certificate.Raw, before unwrapping the
+		// SEQUENCE so it can be operated on
+		if !input.ReadASN1Element(&input, cryptobyte_asn1.SEQUENCE) {
+			return nil, errors.New("x509: malformed certificate")
+		}
+		if !input.ReadASN1(&input, cryptobyte_asn1.SEQUENCE) {
+			return nil, errors.New("x509: malformed certificate")
+		}
+
+		var tbs cryptobyte.String
+		// do the same trick again as above to extract the raw
+		// bytes for Certificate.RawTBSCertificate
+		if !input.ReadASN1Element(&tbs, cryptobyte_asn1.SEQUENCE) {
+			return nil, errors.New("x509: malformed tbs certificate")
+		}
+		if !tbs.ReadASN1(&tbs, cryptobyte_asn1.SEQUENCE) {
+			return nil, errors.New("x509: malformed tbs certificate")
+		}
+
+		if tbs.Empty() {
+			return nil, errors.New("parsed RawTBSCertificate field was empty")
+		}
+
+		return tbs, nil
+	}
+
+	lintRawTBSCert, err := parseTBSCert(&lintCertBytes)
+	if err != nil {
+		return err
+	}
+
+	leafRawTBSCert, err := parseTBSCert(&leafCertBytes)
+	if err != nil {
+		return err
+	}
+
+	if lintRawTBSCert == nil || leafRawTBSCert == nil {
```

Comment:
a) don't directly check for nil here, instead check `len(foo) == 0`
b) no need for this since the anonymous function does the same check on line 673; pick one or the other

Reply:
What is the functional difference between checking for `nil` or `len 0` on a `[]byte`?

Reply:
You're right, the true zero value of `[]byte` is `nil`. But generally the purpose of checks like this is not simply to check for the zero value, but rather to check for obviously invalid inputs, so that more helpful error messages can be displayed and so that we can avoid doing expensive checks if we already know they're going to fail. So a `len(foo) == 0` check also catches the case where the byte slice exists but has an obviously-invalid (zero) length.

Reply:
I can get rid of this because `tbs.Empty()` on line 670 is handling the `len(foo) == 0` case anyways.